opentracing/dapperish/dapper.go

```diff
@@ -17,29 +21,149 @@ type DapperishContextID struct {
+
+	return &DapperishTraceContext{
+		TraceID:   d.TraceID,
+		SpanID:    randomID(),
+		Sampled:   d.Sampled,
+		traceTags: newTags,
 	}, opentracing.Tags{"parent_span_id": d.SpanID}
 }
 
-func (d *DapperishContextID) Serialize() []byte {
+// An implementation of opentracing.TraceContextSource.
+type DapperishTraceContextSource struct{}
+
+func NewDapperishTraceContextSource() *DapperishTraceContextSource {
+	return &DapperishTraceContextSource{}
+}
+
+func (m *DapperishTraceContextSource) NewRootTraceContext() opentracing.TraceContext {
+	return &DapperishTraceContext{
+		TraceID:   randomID(),
+		SpanID:    randomID(),
+		Sampled:   randomID()%1024 == 0,
+		traceTags: make(map[string]string),
+	}
+}
+
+func (d *DapperishTraceContext) SetTraceTag(key, val string) opentracing.TraceContext {
+	d.tagLock.Lock()
+	defer d.tagLock.Unlock()
+
+	d.traceTags[key] = val
+	return d
+}
+
+func (d *DapperishTraceContext) TraceTag(key string) string {
+	d.tagLock.RLock()
+	defer d.tagLock.RUnlock()
+
+	return d.traceTags[key]
+}
+
+func (d *DapperishTraceContextSource) MarshalTraceContextStringMap(
+	ctx opentracing.TraceContext,
+) map[string]string {
+	dctx := ctx.(*DapperishTraceContext)
+	rval := map[string]string{
+		fieldNameTraceId: strconv.FormatInt(dctx.TraceID, 10),
+		fieldNameSpanId:  strconv.FormatInt(dctx.SpanID, 10),
+		fieldNameSampled: strconv.FormatBool(dctx.Sampled),
+	}
+	dctx.tagLock.RLock()
+	for k, v := range dctx.traceTags {
+		rval[fieldNameTagPrefix+k] = v
+	}
+	dctx.tagLock.RUnlock()
+	return rval
+}
+
+func (d *DapperishTraceContextSource) UnmarshalTraceContextStringMap(
+	encoded map[string]string,
+) (opentracing.TraceContext, error) {
+	traceTags := make(map[string]string)
+	requiredFieldCount := 0
+	var traceID, spanID int64
+	var sampled bool
+	var err error
+	for k, v := range encoded {
+		switch k {
+		case fieldNameTraceId:
+			traceID, err = strconv.ParseInt(encoded[fieldNameTraceId], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			requiredFieldCount++
+		case fieldNameSpanId:
+			spanID, err = strconv.ParseInt(encoded[fieldNameSpanId], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			requiredFieldCount++
+		case fieldNameSampled:
+			sampled, err = strconv.ParseBool(encoded[fieldNameSampled])
+			if err != nil {
+				return nil, err
+			}
+			requiredFieldCount++
+		default:
+			if strings.HasPrefix(k, fieldNameTagPrefix) {
+				traceTags[strings.TrimPrefix(k, fieldNameTagPrefix)] = v
+			} else {
+				return nil, fmt.Errorf("Unknown string map field: %v", k)
+			}
+		}
+	}
+	if requiredFieldCount < 3 {
+		return nil, fmt.Errorf("Only found %v of 3 required fields", requiredFieldCount)
+	}
+
+	return &DapperishTraceContext{
+		TraceID:   traceID,
+		SpanID:    spanID,
+		Sampled:   sampled,
+		traceTags: traceTags,
+	}, nil
+}
+
+func (d *DapperishTraceContextSource) MarshalTraceContextBinary(ctx opentracing.TraceContext) []byte {
+	dtc := ctx.(*DapperishTraceContext)
+	// XXX: support tags
```

Comment:
(since this is just a demo impl, I'm leaving this out for now... I handled tags in the string map case.)